pkg/stream/producer.go

```diff
@@ -275,200 +264,236 @@ func (producer *Producer) getStatus() int {
+	// the chSignal is used to signal the two goroutines.
+	// When chSignal is sent by the second goroutine, the first goroutine sends the messages
+	chSignal := make(chan struct{}, producer.options.BatchSize)
 
-						producer.flushUnConfirmedMessages(connectionCloseError, ConnectionClosed)
-						if producer.publishConfirm != nil {
-							close(producer.publishConfirm)
-							producer.publishConfirm = nil
-						}
-						if producer.closeHandler != nil {
-							close(producer.closeHandler)
-							producer.closeHandler = nil
-						}
-						return
-					}
-					producer.mutexPending.Lock()
-					if producer.pendingMessages.size+msg.unCompressedSize >= producer.options.client.getTuneState().
-						requestedMaxFrameSize {
-						producer.sendBufferedMessages()
-					}
+	totalBufferToSend := 0
+	mutexQueue := sync.RWMutex{}
 
-					producer.pendingMessages.size += msg.unCompressedSize
-					producer.pendingMessages.messages = append(producer.pendingMessages.messages, &msg)
-					if len(producer.pendingMessages.messages) >= (producer.options.BatchSize) {
-						producer.sendBufferedMessages()
-					}
-					producer.mutexPending.Unlock()
+	maxFrame := producer.options.client.getTuneState().requestedMaxFrameSize
+
+	// the waitGroup is used to wait for the two goroutines
+	// the first goroutine is the one that sends the messages
+	// the second goroutine is the one that accumulates the messages
+	// the waitGroup is used to wait for the two goroutines
+	waitGroup := sync.WaitGroup{}
+	waitGroup.Add(1)
+	/// send the messages in a batch
+	go func() {
+		waitGroup.Done()
+		for range chSignal {
+			lenB := 0
+			mutexQueue.RLock()
+			lenB = len(queueMessages)
+			mutexQueue.RUnlock()
+
+			if lenB > 0 {
+				mutexQueue.Lock()
+				err := producer.internalBatchSend(queueMessages)
+				if err != nil {
+					logs.LogError("Producer %d, error during batch send: %s", producer.GetID(), err)
 				}
+				queueMessages = queueMessages[:0]
+				totalBufferToSend = initBufferPublishSize
+				mutexQueue.Unlock()
+			}
+		}
+	}()
 
-			case <-ticker.C:
-				producer.mutexPending.Lock()
-				producer.sendBufferedMessages()
-				producer.mutexPending.Unlock()
+	waitGroup.Wait()
+
+	waitGroup.Add(1)
+	/// accumulate the messages in a buffer
+	// the batch messages are sent with the messages that are accumulated in the buffer.
+	// The buffer is reset when the messages are sent.
+	go func() {
+		waitGroup.Done()
+		// Receive the messages from the dynamicSendCh used by Send() function
+		for msg := range producer.dynamicSendCh {
+			toSend := false
+			mutexQueue.Lock()
+			totalBufferToSend += msg.unCompressedSize
+			if totalBufferToSend > maxFrame || len(queueMessages) > producer.options.BatchSize {
+				toSend = true
+			}
+			mutexQueue.Unlock()
+			if toSend {
+				chSignal <- struct{}{}
 			}
+
+			mutexQueue.Lock()
+			queueMessages = append(queueMessages, msg)
+			mutexQueue.Unlock()
+			chSignal <- struct{}{}
```

Comment:
Is not clear to me the double call to `chSignal <- struct{}{}`.
It looks like to me that the client tries to send a message for every message it receives and invalidates the `toSend` logic. Do you agree?
And also I would move the `queueMessages = append(queueMessages, msg)` inside the previous lock, to avoid a double lock.
Also, I don't see the utility of this double routine, given that once you send a signal on `chSignal` the first goroutine will acquire the lock for all the sending period blocking also the second routine, not adding any parallelism. I think it would be less complex and maybe faster to collapse the logic inside one single routine, removing the locks and the coordination channel. What do you think?

Reply:
Changed the way to store the the message